### PR TITLE
Customizable script to allow honor from creature kills

### PR DESCRIPTION
**Pull request by MrSmite**
- See CONF section: CUSTOM SERVER OPTIONS - CREATURE HONOR
- For existing servers, please import: characters_create_creaturehonor_message.sql
- Made a change to CMakeLists.txt which allows inclusion of custom scripts. Re-run CMake if you're not starting fresh.

### CODE REVIEW

File: src/server/shared/Database/Implementation/CharacterDatabase.cpp [hunk 1]
```diff
@@ -608,4 +608,10 @@ void CharacterDatabaseConnection::DoPrepareStatements()
     PrepareStatement(CHAR_UPD_QUEST_TRACK_GM_COMPLETE, "UPDATE quest_tracker SET completed_by_gm = 1 WHERE id = ? AND character_guid = ? ORDER BY quest_accept_time DESC LIMIT 1", CONNECTION_ASYNC);
     PrepareStatement(CHAR_UPD_QUEST_TRACK_COMPLETE_TIME, "UPDATE quest_tracker SET quest_complete_time = NOW() WHERE id = ? AND character_guid = ? ORDER BY quest_accept_time DESC LIMIT 1", CONNECTION_ASYNC);
     PrepareStatement(CHAR_UPD_QUEST_TRACK_ABANDON_TIME, "UPDATE quest_tracker SET quest_abandon_time = NOW() WHERE id = ? AND character_guid = ? ORDER BY quest_accept_time DESC LIMIT 1", CONNECTION_ASYNC);
+
+    // player_creature_honor script
+    PrepareStatement(CHAR_DEL_CREATUREHONOR_MESSAGE, "DELETE FROM creaturehonor_message WHERE guid = ?", CONNECTION_ASYNC);
+    PrepareStatement(CHAR_INS_CREATUREHONOR_MESSAGE, "INSERT IGNORE INTO creaturehonor_message (guid, message_sent) VALUES (?, 0)", CONNECTION_BOTH);
```
Comment (MrSmite) [inline src/server/shared/Database/Implementation/CharacterDatabase.cpp]: Ignore the error if trying to insert a record for a GUID that already exists (duplicate key).

**Comment by MrSmite** [comment]
> @MrSmite I moved SQL logic in Player class, could you please check if jackpoz@5202e92 works the way it's intended to ?

Sure, but I won't be able to get to it until 2015-01-03.

BTW: That's some interesting logic, is that going to be the new standard for loading custom content going forward?

**Comment by jackpoz** [comment]
that's the idea, we'll see in the future if it fits our needs or if it has to be modified

**Comment by MrSmite** [comment]
Sorry I haven't had a chance to test, had to go out of town unexpectedly. Will probably be another few days before I'm back at my machine.

**Comment by MrSmite** [comment]
Got a message the script didn't work... it helps if you actually start the script. Easy enough to fix using direct edits on Github.

**Comment by MrSmite** [comment]
After making the suggested changes in *SendPVPCreditMessage()* regarding the packet and GUIDs, it now says you've died and earned an HK.

In the past it would simply display how much honor you earned, now it says "TestHunter dies. Honorable kill rank: Private (3 honor points)" when TestHunter is me and I've killed an NPC...

**Comment by MrSmite** [comment]
Found a couple minor cosmetic issues that I will fix soon (eg: sending the player's GUID instead of the Creature which results in the wrong message).

**Comment by jackpoz** [comment]
is it with https://github.com/jackpoz/TrinityCore/commit/5202e92 already ? can you apply those changes to this PR ?

**Comment by MrSmite** [comment]
> is it with jackpoz@5202e92 already ? can you apply those changes to this PR ?

Well the big one was the fact that the script wasn't even loading. I edited that directly on Github. The other minor problem is sending the player's GUID causes the chat message to tell the player they died after each NPC kill.

I've made a couple other edits that I'm testing before I push. For instance, the script creates a checksum of the CONF options on load and writes that to the table. If you change a CONF option then the values will differ triggering the mail to be resent. The server admins don't need to muck around in the DB now.

**Comment by MrSmite** [comment]
Update: I made a few changes to the base PR, will work on applying @jackpoz 's changes now.

File: src/server/game/Entities/Player/Player.cpp [hunk 2]
```diff
@@ -7309,6 +7299,31 @@ void Player::SetArenaPoints(uint32 value)
...
+    //
+    //      NOTE: The "HK:" is also displayed as floating text
+
+    // Defaulting victim_guid and victim_rank to zero prevents the client from displaying an
+    //  empty "HK:" floating message if victim has no rank, such as an NPC (script: player_creature_honor)
+    //  or an invalid player (though the second case should NEVER happen)
+
+    WorldPacket data(SMSG_PVP_CREDIT, 4 + 8 + 4);
+    data << honor;
+    data << uint64((victimGuid.IsPlayer() ? victimGuid.GetCounter() : 0)); // must force uint64 or results in wrong message
```
Comment (Shauren) [inline src/server/game/Entities/Player/Player.cpp]: WTF is this? This field is a GUID so why are you only sending a part o it?

**Comment by jackpoz** [comment]
@MrSmite check how many fields in Player class have the m_ prefix and you'll see why I chose not to mix conventions but use the one I found.

**Comment by DDuarte** [comment]
> The most important consistency rules are those that govern naming. The style of a name immediately informs us what sort of thing the named entity is: a type, a variable, a function, a constant, a macro, etc., without requiring us to search for the declaration of that entity. The pattern-matching engine in our brains relies a great deal on these naming rules.

> Naming rules are pretty arbitrary, but we feel that consistency is more important than individual preferences in this area, so regardless of whether you find them sensible or not, the rules are the rules.

in http://google-styleguide.googlecode.com/svn/trunk/cppguide.html

**Comment by jackpoz** [comment]
btw https://github.com/jackpoz/TrinityCore/commit/5202e92 should still apply to latest changes in this PR so a single table stores any character-related custom stuff.
I'm not quite confident about that checksum algorithm, doesn't it give same result if I disable an option and enable another one ?

**Comment by MrSmite** [comment]
> TL;DR While not explicit in our wiki, consistency is better than blindly following the rules.
> A later commit could take all those m_ away.

So then why have I always had to fix variable names in PRs?

**Comment by MrSmite** [comment]
> I'm not quite confident about that checksum algorithm, doesn't it give same result if I disable an option and enable another one ?

Yes, I need to give some weight to differentiate the booleans.

File: src/server/game/Entities/Player/Player.cpp (hunk 2 above)
Comment (MrSmite) [inline src/server/game/Entities/Player/Player.cpp]: > WTF is this? This field is a GUID so why are you only sending a part of it?

Because GetCounter() in the 335 branch sends the GUID as it's found in the database. The other way didn't work.

Comment (Shauren) [inline src/server/game/Entities/Player/Player.cpp]: a full player guid in 335 branch has the same value as GetCounter (because highguid is 0) but that isnt the point, extracting uint32 counter then casting it to uint64 (packet struct requirement) because client expects a full guid doesnt make sense
This line should have been data << (victimGuid.IsPlayer() ? victimGuid : ObjectGuid::Empty)
https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Entities/Object/ObjectGuid.cpp#L71 (operator << ObjectGuid)

Comment (MrSmite) [inline src/server/game/Entities/Player/Player.cpp]: It was like that and it didn't work. It was randomly causing an empty "HK:" message.

Anyway it's moot. After what Nay posted, I've decided to take a break from Trinity for the forseeable future. I don't like the double standards and having to guess which rules to follow and which to ignore for "consistency". In the past I've had to redo a PR for naming convention yet somehow it's OK for you guys?

**Comment by jackpoz** [comment]
It's like learning a new language grammar, there are a lot of rules and even more exceptions but with time and practice it becomes natural.

File: src/server/game/Entities/Player/Player.cpp (hunk 2 above)
Comment (Shauren) [inline src/server/game/Entities/Player/Player.cpp]: Leaving this for future reference
Client handling of PVP_CREDIT
```if (guid != 0)
{
     if (rank >= 0)
    {
        AddCombatLogLine("%s dies, honorable kill Rank: %s (%d Honor Points)");
        AddHonorGainedMessage("HK: %s");
    }
    else
        AddCombatLogLine("%s dies, honorable kill (%d Honor Points)");
}
else
    AddCombatLogLine("You have been awarded %d honor points.");
